# Use Case for Hackathon

### We get a dataset, we have to understand it and get value from it

A school wants to hire us to see whether we are able to understand some problems they believe they are facing. 

Something is going wrong with the performance of the students but they are not able to figure it out and give reasons to it. 

Some parents are very disappointed of the grades of their kids and have complaint continiously to the board of the school. 

Our company is low on budget and we accept the deal even if it is risky, stressful and the money won't cover much of our runaway either but we take what we get. 

## Let's get to it and try to help the School!

### Import dataset and do the first general analysis

In [18]:
import pandas as pd

In [19]:
url='https://drive.google.com/file/d/1ETCwd-hUZ_qhQX0M0OY6OAddVE4bCSOW/view?usp=sharing'
url='https://drive.google.com/uc?id=' + url.split('/')[-2]

df = pd.read_csv(url)
df.head()

gender race/ethnicity parental level of education         lunch  \
0    Male        group A                 high school      standard   
1  Female        group D            some high school  free/reduced   
2    Male        group E                some college  free/reduced   
3    Male        group B                 high school      standard   
4    Male        group E          associate's degree      standard   

  test preparation course  math score  reading score  writing score  \
0               completed          67             67             63   
1                    none          40             29             55   
2                    none          59             60             50   
3                    none          77             78             68   
4               completed          78             73             68   

   id_student  address  Year   Age  
0        1000      NaN  2023  14.0  
1        1001      NaN  2023  17.0  
2        1002      NaN  2023  14.0  
3        1003      NaN  2023  17.0  
4        1004      NaN  2023  16.0

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1010 entries, 0 to 1009
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   gender                       1010 non-null   object 
 1   race/ethnicity               1010 non-null   object 
 2   parental level of education  1010 non-null   object 
 3   lunch                        1010 non-null   object 
 4   test preparation course      1010 non-null   object 
 5   math score                   1010 non-null   int64  
 6   reading score                1010 non-null   int64  
 7   writing score                1010 non-null   int64  
 8   id_student                   1010 non-null   int64  
 9   address                      0 non-null      float64
 10  Year                         1010 non-null   int64  
 11  Age                          943 non-null    float64
dtypes: float64(2), int64(5), object(5)
memory usage: 94.8+ KB


In [28]:
df.describe()

math score  reading score  writing score  address         Year  \
count  1010.000000    1010.000000    1010.000000      0.0  1010.000000   
mean     66.420792      64.930693      70.289109      NaN  2021.104950   
std      15.481187      18.990769      19.125831      NaN     7.681396   
min      13.000000      15.000000      23.000000      NaN  1990.000000   
25%      56.000000      54.000000      58.000000      NaN  2023.000000   
50%      66.000000      68.000000      68.000000      NaN  2023.000000   
75%      77.000000      78.000000      79.000000      NaN  2023.000000   
max     120.000000     100.000000     100.000000      NaN  2023.000000   

              Age  
count  943.000000  
mean    15.483563  
std      1.151707  
min     14.000000  
25%     14.000000  
50%     16.000000  
75%     17.000000  
max     17.000000

Llama la atención el outlier en math score

## What information do we get from this? 

1. Why is student id int? If it is a string. It is not a number

2. We need to inspect all the object columns and check the value counts. What do we discover when we do it? 

3. The distribution of the year column is a bit odd. We need to look more into it. 

4. Do we have duplicates in the dataset?

5. All rows of the column address are missing, maybe we can just delete it

6. The age column has 77 missing values. What should we do with it? 

7. There seems to be values for the math scores higher than 120. Are those outliers? 

### 1. Change type of student id  

In [21]:
#está cogiendo id como un número y no como variable
df['id_student']=df['id_student'].astype(str)

### 2. Explore the columns that are objects (string)

In [22]:
object_columns = df.select_dtypes(include='object')
object_columns.head()

gender race/ethnicity parental level of education         lunch  \
0    Male        group A                 high school      standard   
1  Female        group D            some high school  free/reduced   
2    Male        group E                some college  free/reduced   
3    Male        group B                 high school      standard   
4    Male        group E          associate's degree      standard   

  test preparation course id_student  
0               completed       1000  
1                    none       1001  
2                    none       1002  
3                    none       1003  
4               completed       1004

### 3. Seems that the column gender has a lot of versions we are going to have to fix this

In [23]:
df['gender'].unique()

array(['Male', 'Female', 'MALE', 'FEMALE', 'male', 'female'], dtype=object)

In [24]:
df['gender'] = df['gender'].str.upper()
df['gender'].unique()

array(['MALE', 'FEMALE'], dtype=object)

### 4. What may be happening with the year? 

In [26]:
df['Year'].describe()

count    1010.000000
mean     2021.104950
std         7.681396
min      1990.000000
25%      2023.000000
50%      2023.000000
75%      2023.000000
max      2023.000000
Name: Year, dtype: float64

In [29]:
df['Year'].value_counts()

Year
2023    952
1990     58
Name: count, dtype: int64

#hay valores de año que no tienen sentido pero nos centraremos en otras variables. Podríasmos reemplazar con la media o ignorarlo

### 5. Do we have any duplicates? 

In [30]:
df_1=df.drop_duplicates()
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   gender                       1000 non-null   object 
 1   race/ethnicity               1000 non-null   object 
 2   parental level of education  1000 non-null   object 
 3   lunch                        1000 non-null   object 
 4   test preparation course      1000 non-null   object 
 5   math score                   1000 non-null   int64  
 6   reading score                1000 non-null   int64  
 7   writing score                1000 non-null   int64  
 8   id_student                   1000 non-null   object 
 9   address                      0 non-null      float64
 10  Year                         1000 non-null   int64  
 11  Age                          933 non-null    float64
dtypes: float64(2), int64(4), object(6)
memory usage: 101.6+ KB


In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1010 entries, 0 to 1009
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   gender                       1010 non-null   object 
 1   race/ethnicity               1010 non-null   object 
 2   parental level of education  1010 non-null   object 
 3   lunch                        1010 non-null   object 
 4   test preparation course      1010 non-null   object 
 5   math score                   1010 non-null   int64  
 6   reading score                1010 non-null   int64  
 7   writing score                1010 non-null   int64  
 8   id_student                   1010 non-null   object 
 9   address                      0 non-null      float64
 10  Year                         1010 non-null   int64  
 11  Age                          943 non-null    float64
dtypes: float64(2), int64(4), object(6)
memory usage: 94.8+ KB


Vemos entre los dos datasets que la diferencia es de 10 duplicados

### 6.1. Missing values: Remember the column address

In [ ]:
#la columna no contiene datos por lo tanto la borramos
df_1=df_1.drop('address',axis=1)

In [34]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 0 to 999
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   gender                       1000 non-null   object 
 1   race/ethnicity               1000 non-null   object 
 2   parental level of education  1000 non-null   object 
 3   lunch                        1000 non-null   object 
 4   test preparation course      1000 non-null   object 
 5   math score                   1000 non-null   int64  
 6   reading score                1000 non-null   int64  
 7   writing score                1000 non-null   int64  
 8   id_student                   1000 non-null   object 
 9   Year                         1000 non-null   int64  
 10  Age                          933 non-null    float64
dtypes: float64(1), int64(4), object(6)
memory usage: 126.0+ KB


### 6.2. What about the missing values of the column Age? 

Hay 67 valores nulos en la columna (visto en el apartado anterior)

### 7. What is going on with the math scores?

In [36]:
#hemos visto que la puntuación máxima está por encima de 100
df_1.loc[df_1['math score'] > 100]

gender race/ethnicity parental level of education     lunch  \
11  FEMALE        group D                 high school  standard   

   test preparation course  math score  reading score  writing score  \
11               completed         120             87            100   

   id_student  Year   Age  
11       1011  2023  16.0

Solo hay un otulier

## Do some initial plots? EDA

In [ ]:
#entregar hasta antes de esto

In [47]:
df_2=df_1.groupby(by=['test preparation course'])[('math score')].mean()
df_2.head()

test preparation course
completed    69.805970
none         64.738346
Name: math score, dtype: float64

Los alumnos que no han completado el curso preparatorio obtienen peores resultados en promedio

### What do you find interesting?  

- To me the reading score is really intriguing --> We need to really work on understanding what is going on
- The writing score is also characteristic but in principle I am not that worried about it. 

#### We know for a fact that the id_student and the year will not be relevant for any analysis as they are the same value for the whole dataset

Maybe we can delete them from now on

#### What do we care about? The grades are the main problem. Maybe they vary based on any of the other variables

I think we can detect the biggest discrepancies at 3-4 levels: 

- Parental level of education 

- Test preparation course 

- Mybe lunch or gender (I am pretty sure the lunch should not have that much impact but let's still check it)

Let's go deeper

 



### 2-variable Analysis with Lunch column 

### 2-variable Analysis with gender column 

### 2-variable Analysis with Test preparation course column 

### What is next?

- Get the story together
- Improve your code
- Get some statistics of the differences you want to show 
- Work on the plots to show the data nicely 
- Get the presentation ready 
- Show what you found to the client